<a href="https://colab.research.google.com/github/slouie88/pandas-practice/blob/master/Python_pandas_Practice_Problems_UC_Berkeley_MIDS_notebook.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Python pandas Practice Problems

Welcome to datascience@berkeley's collection of Python pandas practice problems.

To use this notebook, go to `File` --> `Save a copy in Drive` in the menu bar above to create a copy of this notebook on your computer. A new tab will open with the copy and you will be the owner. Your version will not be visible to anyone else.

A few notes to get you started:

* The space provided for code is not a line limit or hint. Add and use as many lines of code or additional cells as necessary.
* Lines preceded with '#' are comments and provide additional explanation but will not affect written code or execution.
* Solutions can be found in the complementary [Python pandas Practice Problems for Beginners - Solutions notebook](https://colab.research.google.com/drive/1f8PTrhgJR093fBth5Ka23KCWTIGbgPds?usp=sharing).
* Additional resources for the concepts used in each problem can be found at [the blog post here](https://ischoolonline.berkeley.edu/blog/python-pandas-practice-problems/).


In [1]:
# Import statements go here

import pandas as pd
import statsmodels.api as sm

## Importing Data and Making a DataFrame
The statsmodels package (installed in the code cell above) includes built-in datasets. Execute the code below to download data from the [American National Election Studies of 1996](https://www.statsmodels.org/dev/datasets/generated/anes96.html) and print a detailed description of the schema.

The next cell extracts the `Dataset` object from the submodule and saves the `DataFrame` to the variable `df`. In the questions that follow, use `df` when referencing the dataset.

In [2]:
anes96 = sm.datasets.anes96
print(anes96.NOTE)

::

    Number of observations - 944
    Number of variables - 10

    Variables name definitions::

            popul - Census place population in 1000s
            TVnews - Number of times per week that respondent watches TV news.
            PID - Party identification of respondent.
                0 - Strong Democrat
                1 - Weak Democrat
                2 - Independent-Democrat
                3 - Independent-Indpendent
                4 - Independent-Republican
                5 - Weak Republican
                6 - Strong Republican
            age : Age of respondent.
            educ - Education level of respondent
                1 - 1-8 grades
                2 - Some high school
                3 - High school graduate
                4 - Some college
                5 - College degree
                6 - Master's degree
                7 - PhD
            income - Income of household
                1  - None or less than $2,999
                2  - $3,000-$4,9

In [3]:
dataset_anes96 = anes96.load_pandas()
df = dataset_anes96.data

## 1. DataFrame Basic Properties Exercise

Our DataFrame (`df`) contains data on registered voters in the United States, including demographic information and political preference. Using `pandas`, print the first 5 rows of the DataFrame to get a sense of what the data looks like. Next, answer the following questions:


*   How many observations are in the DataFrame?
    - 944 observations
*   How many variables are measured (how many columns)?
    - 11 variables
*   What is the age of the youngest person in the data? The oldest?
    - Youngest age: 19yrs old
    - Oldest age: 91 yrs old
*   How many days a week does the average respondent watch TV news (round to the nearest tenth)?
    - 3.7 times
*   Check for missing values. Are there any?
    - None






In [5]:
# Your code here
df.head()

popul  TVnews  selfLR  ClinLR  DoleLR  PID   age  educ  income  vote  \
0    0.0     7.0     7.0     1.0     6.0  6.0  36.0   3.0     1.0   1.0   
1  190.0     1.0     3.0     3.0     5.0  1.0  20.0   4.0     1.0   0.0   
2   31.0     7.0     2.0     2.0     6.0  1.0  24.0   6.0     1.0   0.0   
3   83.0     4.0     3.0     4.0     5.0  1.0  28.0   6.0     1.0   0.0   
4  640.0     7.0     5.0     6.0     4.0  0.0  68.0   6.0     1.0   0.0   

   logpopul  
0 -2.302585  
1  5.247550  
2  3.437208  
3  4.420045  
4  6.461624

In [6]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 944 entries, 0 to 943
Data columns (total 11 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   popul     944 non-null    float64
 1   TVnews    944 non-null    float64
 2   selfLR    944 non-null    float64
 3   ClinLR    944 non-null    float64
 4   DoleLR    944 non-null    float64
 5   PID       944 non-null    float64
 6   age       944 non-null    float64
 7   educ      944 non-null    float64
 8   income    944 non-null    float64
 9   vote      944 non-null    float64
 10  logpopul  944 non-null    float64
dtypes: float64(11)
memory usage: 81.2 KB


In [9]:
youngest = df["age"].min()
oldest = df["age"].max()
print(f"Age of the youngest person observed in the dataset: {youngest}")
print(f"Age of the oldest person observed in the dataset: {oldest}")

Age of the youngest person observed in the dataset: 19.0
Age of the oldest person observed in the dataset: 91.0


In [10]:
average_tv_per_week = df["TVnews"].mean()
print(f"The average person watches TV approximately {round(average_tv_per_week, 1)} times per week")

The average person watches TV approximately 3.7 times per week


## 2. Data Processing Exercise

We want to adjust the dataset for our use. Do the following:


*   Rename the `educ` column `education`.
*   Create a new column called `party` based on each respondent's answer to `PID`. `party` should equal `Democrat` if the respondent selected either Strong Democrat or Weak Democrat. `party` will equal `Republican` if the respondent selected Strong or Weak Republican for `PID` and `Independent` if they selected anything else.
*   Create a new column called `age_group` that buckets respondents into the following categories based on their `age`: 18-24, 25-34, 35-44, 45-54, 55-64, and 65 and over.



In [14]:
# Your code here
df.rename(columns={"educ": "education"}, inplace=True)
df.head()

popul  TVnews  selfLR  ClinLR  DoleLR  PID   age  education  income  vote  \
0    0.0     7.0     7.0     1.0     6.0  6.0  36.0        3.0     1.0   1.0   
1  190.0     1.0     3.0     3.0     5.0  1.0  20.0        4.0     1.0   0.0   
2   31.0     7.0     2.0     2.0     6.0  1.0  24.0        6.0     1.0   0.0   
3   83.0     4.0     3.0     4.0     5.0  1.0  28.0        6.0     1.0   0.0   
4  640.0     7.0     5.0     6.0     4.0  0.0  68.0        6.0     1.0   0.0   

   logpopul  
0 -2.302585  
1  5.247550  
2  3.437208  
3  4.420045  
4  6.461624

In [15]:
party = []
for i in df["PID"]:
    if i == 0 or i == 1:
        party.append("Democrat")
    elif i == 5 or i == 6:
        party.append("Republican")
    else:
        party.append("Independent")
df["party"] = party
df.head()

popul  TVnews  selfLR  ClinLR  DoleLR  PID   age  education  income  vote  \
0    0.0     7.0     7.0     1.0     6.0  6.0  36.0        3.0     1.0   1.0   
1  190.0     1.0     3.0     3.0     5.0  1.0  20.0        4.0     1.0   0.0   
2   31.0     7.0     2.0     2.0     6.0  1.0  24.0        6.0     1.0   0.0   
3   83.0     4.0     3.0     4.0     5.0  1.0  28.0        6.0     1.0   0.0   
4  640.0     7.0     5.0     6.0     4.0  0.0  68.0        6.0     1.0   0.0   

   logpopul       party  
0 -2.302585  Republican  
1  5.247550    Democrat  
2  3.437208    Democrat  
3  4.420045    Democrat  
4  6.461624    Democrat

In [16]:
age_group = []
for age in df["age"]:
    if age >= 18 and age < 25:
        age_group.append("18-24")
    elif age >= 25 and age < 35:
        age_group.append("25-35")
    elif age >= 35 and age < 45:
        age_group.append("35-45")
    elif age >= 45 and age < 55:
        age_group.append("45-55")
    elif age >= 55 and age < 65:
        age_group.append("55-65")
    else:
        age_group.append("65 and over")
df["age_group"] = age_group
df.head()

popul  TVnews  selfLR  ClinLR  DoleLR  PID   age  education  income  vote  \
0    0.0     7.0     7.0     1.0     6.0  6.0  36.0        3.0     1.0   1.0   
1  190.0     1.0     3.0     3.0     5.0  1.0  20.0        4.0     1.0   0.0   
2   31.0     7.0     2.0     2.0     6.0  1.0  24.0        6.0     1.0   0.0   
3   83.0     4.0     3.0     4.0     5.0  1.0  28.0        6.0     1.0   0.0   
4  640.0     7.0     5.0     6.0     4.0  0.0  68.0        6.0     1.0   0.0   

   logpopul       party    age_group  
0 -2.302585  Republican        35-45  
1  5.247550    Democrat        18-24  
2  3.437208    Democrat        18-24  
3  4.420045    Democrat        25-35  
4  6.461624    Democrat  65 and over

## 3. Filtering Data Exercise

Use the filtering method to find all the respondents who have the impression that Bill Clinton is moderate or conservative (`ClinLR` equals 4 or higher). How many respondents are in this subset?

Among these respondents, how many have a household income less than $50,000 and attended at least some college?

In [ ]:
# Your code here

## 4. Calculating From Data Exercise

For each of the below match-ups, choose the group that is more likely to vote for Bill Clinton. You can calculate this using the percentage of each group that intends to vote for Clinton (`vote`). Which match-up was the closest? Which had the biggest difference?

Another way to think about this: Given that a respondent is a Democrat, there is a ____ percent chance they will vote for Clinton. How does this value change if the respondent is a Republican?

*   Democrats or Republicans
*   People younger than 44 or People 44 and older
*   People who watch TV news at least 6 days a week or People who watch TV news less than 3 days a week
*   People who live somewhere with a population greater than the average respondent or People who live in a place with a population equal to or less than the average respondent


In [ ]:
# Your code here

## 5. Grouping Data Exercise

Use the `groupby()` method to bucket respondents by `age_group`. Which age group is the most conservative? Which watches TV news the least?

Next, calculate 5 percentile groups based on income. Group the dataset by these percentiles. Which income bracket is the most liberal? Which is the most conservative? The oldest? Highest educated?

In [ ]:
# Your code here

## 6. Voting Across the Aisle

We are interested in learning more about respondents who's political views differ strongly from the candidate they expect to vote for. Using `selfLR`, `vote`, `ClinLR`, and `DoleLR`, work through the following questions. Your interpretation may differ from the answer key.

*   What is the largest recorded difference between a respondent's political leaning and their impression of their intended candidate's political leaning?
*   How many respondents exhibit a difference of that magnitude?
*   Make a separate DataFrame called `sway` that only includes voters who exhibit a difference greater than |3|.
*   Among those in `sway`, are respondents more likely to be voting for a candidate more conservative or more liberal than their own political leaning?
*   In `sway`, which candidate is the more popular choice?



In [ ]:
# Your code here

# BSD 3-Clause License

*Copyright (c) 2022, UC Berkeley School of Information*

*All rights reserved.*

*Redistribution and use in source and binary forms, with or without modification, are permitted provided that the following conditions are met:*

*1. Redistributions of source code must retain the above copyright notice, this list of conditions and the following disclaimer.*

*2. Redistributions in binary form must reproduce the above copyright notice, this list of conditions and the following disclaimer in the documentation and/or other materials provided with the distribution.*

*3. Neither the name of the copyright holder nor the names of its contributors may be used to endorse or promote products derived from this software without specific prior written permission.*

*THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.*